README
App-TimeClock

This is a simple reporting utility for timeclock, which is an Emacs
time tracking package.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc App::TimeClock

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-TimeClock

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/App-TimeClock

    CPAN Ratings
        http://cpanratings.perl.org/d/App-TimeClock

    Search CPAN
        http://search.cpan.org/dist/App-TimeClock/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Søren Lund

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 dated June, 1991 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.



Changes
Revision history for App-TimeClock

0.01    2012-03-15 22:20
        First version, released on an unsuspecting world.

0.02    2012-03-16 11:07
        Added description of .timelogrc configuration file.
        Fixed a few links in the documentation.

0.03    2012-03-16 16:02
        ConsolePrinter: Added a blank line before printing the header.
        timeclock.pl: Added --version option that will print the version number.
        Switched to GPL v3.
        Updated documentation.


lib/App/TimeClock/CsvPrinter.pm
package App::TimeClock::CsvPrinter;

our @ISA = qw(App::TimeClock::PrinterInterface);

use POSIX qw(strftime);

=head1 NAME

App::TimeClock::CsvPrinter

=head1 DESCRIPTION 

Implements the L<App::TimeClock::PrinterInterface>. Will print total
for each day in a comma separated format.

=head1 METHODS

=over

=cut

=item print_header()

There's no header when print CSV. This is an empty sub.

=cut
sub print_header {};

=item print_day()

Prints totals for each day. Five fields are printed: week day, date,
start time, end time and total hours worked. Example:

 "Mon","2012/03/12","08:21:16","16:05:31",7.732222

=cut
sub print_day {
    my ($self, $date, $start, $end, $work, %projects) = (@_);
    my ($year, $mon, $mday) = split(/\//, $date);
    my $wday = substr(strftime("%a", 0, 0, 0, $mday, $mon-1, $year-1900),0,3);

    printf '"%s","%s","%s","%s",%f' . "\n",$wday, $date, $start, $end, $work;
};

=item print_footer()

There's no footer when print CSV. This is an empty sub.

=cut
sub print_footer {};
1;

=back

=for text
=encoding utf-8
=end

=head1 AUTHOR

Søren Lund, C<< <soren at lund.org> >>

=head1 SEE ALSO

L<timeclock.pl>

=head1 COPYRIGHT

Copyright (C) 2012 Søren Lund

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 dated June, 1991 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


lib/App/TimeClock/DailyReport.pm
package App::TimeClock::DailyReport;

=head1 NAME

App::TimeClock::DailyReport

=head1 DESCRIPTION

Can parse the timelog and generate a report using an instance of a
L<App::TimeClock::PrinterInterface>.

=head2 METHODS

=over

=item new($timelog, $printer)

Initializes a new L<App::TimeClock::DailyReport> object.

Two parameters are required:

=over

=item B<$timelog>

Must point to a timelog file. Will die if not.

=item B<$printer>

An object derived from L<App::TimeClock::PrinterInterface>. Will die if not.

=back

=cut
sub new {
    my $class = shift;
    my $self = {
        timelog => shift,
        printer => shift,
    };
    die "timelog ($self->{timelog}) does not exist" unless -f $self->{timelog} and -r $self->{timelog};
    die "printer is not a PrinterInterface" unless $self->{printer}->isa("App::TimeClock::PrinterInterface");
    bless $self, $class;
};

=item execute()

Opens the timelog file starts parsing it, looping over each day and
calling print_day() for each.

=cut
sub execute {
    use Time::Local;
    use POSIX qw(difftime);

    my $self = shift;

    open FILE, '<', $self->{timelog} or die "$!\n";

    my %projects;
    my ($current_project, $current_date, $work, $work_total, $start, $end);
    my ($start_time, $end_time);
    my ($work_year_to_date, $day_count) = (0,0);

    $current_date = "";
    $work_total = 0;

    $self->{printer}->print_header;

    while(<FILE>) {

        # Split the line, it should contain:
        #
        # - state is either 'i' - check in or 'o' - check out.
        # - date is formatted as YYYY/MM//DD
        # - time is formatted as HH:MM:SS
        # - project is then name of the project/task and is only required when checking in.
        #
        my ($state, $date, $time, $project) = split(/ /, $_, 4);

        if ($current_project) {
            chomp($current_project);
            $current_project=~tr/\r//d;
        }

        chomp($date);

        chomp($time); $time=~tr/\r//d;

        my ($hours, $min, $sec ) = split(/:/, $time);

        my ($year, $mon, $mday) = split(/\//, $date);

        if ($state eq 'i') {
            if (!length($current_date)) {
                $current_date = $date;
                $start_time = $time;
            } elsif (!($current_date eq $date)) {
                $self->{printer}->print_day($current_date, $start_time, $end_time, $work_total, %projects);

                $work_year_to_date += $work_total;
                $day_count++;

                $work_total = 0;
                $current_date = $date;
                $start_time = $time;
                %projects = ();
                $end_time = "";
            }

            $current_project = $project;
            $start = timelocal($sec,$min,$hours,$mday,$mon-1,$year);

        } elsif ($state eq 'o') {
            $end = timelocal($sec,$min,$hours,$mday,$mon-1,$year);
            $work = difftime($end, $start) / 60 / 60;
            $work_total += $work;
            $end_time = $time;
            $projects{$current_project} += $work;
        }
    }

    if (length($current_date)) {
	$self->{printer}->print_day($current_date, $start_time, $end_time, $work_total, %projects);
	$work_year_to_date += $work_total;
	$day_count++;
    }

    $self->{printer}->print_footer($work_year_to_date, $day_count);
};
1;

=back

=for text
=encoding utf-8
=end

=head1 AUTHOR

Søren Lund, C<< <soren at lund.org> >>

=head1 SEE ALSO

L<timeclock.pl>

=head1 COPYRIGHT

Copyright (C) 2012 Søren Lund

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 dated June, 1991 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


lib/App/TimeClock/HtmlPrinter.pm
package App::TimeClock::HtmlPrinter;

our @ISA = qw(App::TimeClock::PrinterInterface);

use POSIX qw(strftime);

=head1 NAME

App::TimeClock::HtmlPrinter

=head1 DESCRIPTION

Implements the L<App::TimeClock::PrinterInterface>. Will print a
(simple) HTML report with embedded (CSS) styling.

=head1 METHODS

=over

=cut

=item print_header()

Prints a standard HTML header, i.e. DOCTYPE followed by an opening html
tag and a head tag. The head will contain a title with todays date and
an embedded (css) style. A body tag will open and an h1 header with
todays day will be printed. The body and html tags will be closed in the
L</print_footer()> method.

=cut
sub print_header {
    my $title = "Daily Report " . localtime();
    print << "EOD";
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><meta http-equiv='Content-Type' content='text/html;charset=utf-8'/><title>$title</title>
<style type='text/css'>
html, body { margin: 0; padding: 0; border: 0; }
body { background:#fff; color:#000; margin: 1em 0 0 1em; }
table { border-collapse: collapse; width: 80ex; }
th, td { margin:0; border:1px solid #000;padding:0.2em; }
caption { margin-top: 1em; }
tr, caption { text-align:left; }
.totals { border: 2px solid #000; background: #9cf; width: 40ex; padding: 0.5em;}
.totals, caption { font-size: 110%; font-weight: bold; }
th { background: #58b; color: #fff; }
tr:nth-child(even) td { background: #dfe; }
tr:nth-child(odd) td { background: #def; }
th.N, td.N { width: 10%; text-align: right; }
</style></head><body><h1>$title</h1>
EOD

};

=item print_day()

Prints all activities for a day including a total. Is printed in a
standard HTML table.

=cut
sub print_day {
    my ($self, $date, $start, $end, $work, %projects) = (@_);
    my ($year, $mon, $mday) = split(/\//, $date);
    my $wday = substr(strftime("%a", 0, 0, 0, $mday, $mon-1, $year-1900),0,3);

    printf "<table><caption>%3s %s (%s - %s)</caption>\n", $wday, $date, $start, $end;
    printf "<tr><th>Total Daily Hours</th><th class='N'>%5.2f</th></tr>\n", $work;

    foreach my $k (sort keys %projects) {
        printf("<tr><td>%-60s</td><td class='N'>%5.2f</td></tr>\n",
               $k, $projects{$k});
    }
    print "</table>\n";
};

=item print_footer()

Prints the total number of hours worked and the daily average and closes
the body and html tags.

=cut
sub print_footer {
    my ($self, $work_year_to_date, $day_count) = (@_);
    print "<p class='totals'>";
    printf "TOTAL = %.2f<br/>", $work_year_to_date;
    printf "PERIOD = %d days<br/>", $day_count;
    printf "AVG. DAY = %.2f<br/>", $work_year_to_date / $day_count;
    print "</p>\n";
    print "</body></html>\n";
};
1;

=back

=for text
=encoding utf-8
=end

=head1 AUTHOR

Søren Lund, C<< <soren at lund.org> >>

=head1 SEE ALSO

L<timeclock.pl>

=head1 COPYRIGHT

Copyright (C) 2012 Søren Lund

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 dated June, 1991 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


lib/App/TimeClock/PrinterInterface.pm
package App::TimeClock::PrinterInterface;

=head1 NAME

App::TimeClock::PrinterInterface

=head1 DESCRIPTION

Interface class. All printer objects given to
L<App::TimeClock::DailyReport> constructor must be derived from
PrinterInterface.

=head1 SYNOPSIS

 package App::TimeClock::MyPrinter;
 our @ISA = qw(App::TimeClock::PrinterInterface);
 ...
 sub print_header {
     ...
 }
 sub print_day {
     ...
 }
 sub print_footer {
     ...
 }

=head1 METHODS

=over

=cut

=item new()

Creates a new object.

=cut
sub new {
    bless { }, shift;
}

=item print_header()

Called once at the start of a report.

=cut
sub print_header { shift->_must_implement; };

=item print_day()

Called for each day in the report.

=cut
sub print_day { shift->_must_implement; };

=item print_footer()

Called once at the end of a report.

=cut
sub print_footer { shift->_must_implement; };

sub _must_implement {
    (my $name = (caller(1))[3]) =~ s/^.*:://;
    my ($filename, $line) = (caller(0))[1..2];
    die "You must implement $name() method at $filename line $line";
}
1;

=back

=for text
=encoding utf-8
=end

=head1 AUTHOR

Søren Lund, C<< <soren at lund.org> >>

=head1 SEE ALSO

L<timeclock.pl>

=head1 COPYRIGHT

Copyright (C) 2012 Søren Lund

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 dated June, 1991 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


lib/App/TimeClock/ConsolePrinter.pm
package App::TimeClock::ConsolePrinter;

our @ISA = qw(App::TimeClock::PrinterInterface);

use POSIX qw(strftime);

our $hrline =  '+' . ('-' x 62) . '+' . ('-' x 7) . '+';

=head1 NAME

App::TimeClock::ConsolePrinter

=head1 DESCRIPTION

Implements the L<App::TimeClock::PrinterInterface>. Will print a simple ASCII
format. Suitable for using in a console/terminal.

=head1 METHODS

=over

=cut

=item print_header()

Prints a header including todays date. The header is indented to be
centered above the tables printed by L</print_day()>. Example:

                 =====================================
                 Daily Report Wed Mar 14 13:39:06 2012
                 =====================================

=cut
sub print_header {
    my $ident = ' ' x 17;
    print "\n";
    print "${ident}=====================================\n";
    print "${ident}Daily Report " . localtime() . "\n";
    print "${ident}=====================================\n\n";
};

=item print_day()

Prints all activities for a day including a total. Is printed in a ACSII
table. Example:

 * Mon 2012/03/12 (08:21:16 - 16:05:31) *
 +--------------------------------------------------------------+-------+
 | Total Daily Hours                                            |  7.73 |
 +--------------------------------------------------------------+-------+
 | Lunch                                                        |  0.57 |
 +--------------------------------------------------------------+-------+
 | MyProject:Estimation                                         |  2.90 |
 +--------------------------------------------------------------+-------+
 | AnotherProject:BugFixing                                     |  4.26 |
 +--------------------------------------------------------------+-------+

=cut
sub print_day {
    my ($self, $date, $start, $end, $work, %projects) = (@_);
    my ($year, $mon, $mday) = split(/\//, $date);
    my $wday = substr(strftime("%a", 0, 0, 0, $mday, $mon-1, $year-1900),0,3);

    printf("* %3s %s (%s - %s) *\n", $wday, $date, $start, $end);
    print("$hrline\n");
    printf("| Total Daily Hours                                            | %5.2f |\n", $work);
    print("$hrline\n");

    foreach my $k (sort keys %projects) {
        printf("| %-60s | %5.2f |\n",$k, $projects{$k});
        print("$hrline\n");
    }
    printf("\n");
};

=item print_footer()

Prints the total number of hours worked and the daily
average. Example:

 TOTAL = 291.72 hours
 PERIOD = 42 days
 AVERAGE = 7.95 hours/day

=cut
sub print_footer {
    my ($self, $work_year_to_date, $day_count) = (@_);
    printf "TOTAL = %.2f hours\n", $work_year_to_date;
    printf "PERIOD = %d days\n", $day_count;
    printf "AVERAGE = %.2f hours/day\n", $work_year_to_date / $day_count;
};
1;

=back

=for text
=encoding utf-8
=end

=head1 AUTHOR

Søren Lund, C<< <soren at lund.org> >>

=head1 SEE ALSO

L<timeclock.pl>

=head1 COPYRIGHT

Copyright (C) 2012 Søren Lund

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 dated June, 1991 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
